elasticdl/pkg/ps/server.go

```diff
@@ -63,23 +71,53 @@ func createMasterClient(masterAddr string) *MasterClient {
 }
 
 // NewServer creates a Server instance
-func NewServer(ID int, optType string, optArgs string, masterAddr string, evaluationSteps int) *Server {
+func NewServer(ID int, optType string, optArgs string, masterAddr string,
+	evaluationStep int, checkpointDirForInit string,
+	checkpointDir string, checkpointStep int, keepCheckpointMax int, numPsPods int) *Server {
 	var ps Server
-	ps.Model = NewModel()
+	if checkpointDirForInit != "" {
+		var err error
+		ps.Model, err = LoadModelFromCheckpoint(checkpointDirForInit, ID, numPsPods)
+		ps.Model.Initialized = true
+		if err != nil {
+			log.Fatalf("failed to load from checkpoint: %v", err)
+		}
+	} else {
+		ps.Model = NewModel()
+	}
+
 	var err error
 	ps.Opt, err = NewOptimizer(optType, optArgs)
 	if err != nil {
 		log.Fatalf("failed to create PS server: %v", err)
 	}
 	ps.ID = ID
 	ps.masterClient = createMasterClient(masterAddr)
-	ps.evaluationSteps = int32(evaluationSteps)
+	ps.evaluationStep = evaluationStep
+	ps.checkpointDirForInit = checkpointDirForInit
+	ps.checkpointDir = checkpointDir
+	ps.checkpointStep = checkpointStep
+	ps.keepCheckpointMax = keepCheckpointMax
+	ps.numPsPods = numPsPods
 	return &ps
 }
 
-func (s *Server) reportModelVersionIfNeeded(modelVersion int32) {
-	if s.evaluationSteps > 0 && modelVersion%s.evaluationSteps == 0 && s.masterClient != nil {
-		s.masterClient.reportVersion(modelVersion)
+func (s *Server) reportModelVersionIfNeeded(modelVersion int) {
+	if s.evaluationStep > 0 && modelVersion%s.evaluationStep == 0 && s.masterClient != nil {
+		s.masterClient.reportVersion(int32(modelVersion))
+	}
+}
+
+func (s *Server) saveCheckpointIfNeeded(modelVersion int) {
+	if s.checkpointDir != "" && s.checkpointStep != 0 && modelVersion%s.checkpointStep == 0 {
+		checkpointVersionDir := path.Join(s.checkpointDir, fmt.Sprintf("version-%d", modelVersion))
+		s.savedCheckpointDirs = append(s.savedCheckpointDirs, checkpointVersionDir)
+		SaveModelToCheckpoint(checkpointVersionDir, s.Model, s.ID, s.numPsPods)
+		if len(s.savedCheckpointDirs) > s.keepCheckpointMax {
+			deletedDir := s.savedCheckpointDirs[0]
+			s.savedCheckpointDirs = s.savedCheckpointDirs[1:]
+			os.RemoveAll(deletedDir)
```

Comment:
The `deleteDir` may have been deleted by other ps instance. We can only allow the first ps instance to delete the old checkpoint directories.

Reply:
Good idea. Fix it.